substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+		);
+
+		// ensure staker is a nominator
+		let status = T::CoreStaking::status(new_delegatee)?;
+		match status {
+			StakerStatus::Nominator(_) => (),
+			_ => return Err(Error::<T>::InvalidDelegation.into()),
+		}
+
+		<DelegateeMigration<T>>::insert(&new_delegatee, &proxy_delegator);
+		let stake = T::CoreStaking::stake(new_delegatee)?;
+
+		// unlock funds from staker
+		T::CoreStaking::release_all(new_delegatee);
+
+		// try transferring the staked amount. This should never fail but if it does, it indicates
+		// bad state and we abort.
+		T::Currency::transfer(
+			new_delegatee,
+			proxy_delegator,
+			stake.total,
+			Preservation::Expendable,
+		)
+		.map_err(|_| Error::<T>::BadState)?;
+
+		// delegate from new delegator to staker.
+		// todo(ank4n) : inline this fn and propagate payee to core staking..
+		Self::accept_delegations(new_delegatee, payee)?;
+
+		Self::delegate(proxy_delegator, new_delegatee, stake.total)?;
+		Self::bond_all(new_delegatee)
+	}
+
+	fn block_delegations(delegatee: &Self::AccountId) -> DispatchResult {
+		let mut register = <Delegatees<T>>::get(delegatee).ok_or(Error::<T>::NotDelegatee)?;
+		register.blocked = true;
+		<Delegatees<T>>::insert(delegatee, register);
+
+		Ok(())
+	}
+
+	fn unblock_delegations(delegatee: &Self::AccountId) -> DispatchResult {
+		let mut register = <Delegatees<T>>::get(delegatee).ok_or(Error::<T>::NotDelegatee)?;
+		register.blocked = false;
+		<Delegatees<T>>::insert(delegatee, register);
+
+		Ok(())
+	}
+
+	fn kill_delegatee(_delegatee: &Self::AccountId) -> DispatchResult {
+		todo!()
+	}
+
+	fn bond_all(who: &Self::AccountId) -> DispatchResult {
+		let delegatee = <Delegatees<T>>::get(who).ok_or(Error::<T>::NotDelegatee)?;
+		let amount_to_bond = delegatee.unbonded_balance();
+
+		match T::CoreStaking::stake(who) {
+			// already bonded
+			Ok(_) => T::CoreStaking::bond_extra(who, amount_to_bond),
+			// first bond
+			Err(_) => T::CoreStaking::bond(who, amount_to_bond, &delegatee.payee),
+		}
+	}
+
+	#[transactional]
+	fn withdraw(
+		delegator: &Self::AccountId,
+		delegatee: &Self::AccountId,
+		value: Self::Balance,
+		num_slashing_spans: u32,
+	) -> DispatchResult {
+		// check how much is already unbonded
+		let delegation_register =
+			<Delegatees<T>>::get(delegatee).ok_or(Error::<T>::NotDelegatee)?;
+		let unbonded_balance = delegation_register.unbonded_balance();
+
+		if unbonded_balance < value {
+			// fixme(ank4n) handle killing of stash
+			let amount_to_withdraw = value.saturating_sub(unbonded_balance);
+			let _stash_killed: bool =
+				T::CoreStaking::withdraw_exact(delegatee, amount_to_withdraw, num_slashing_spans)
+					.map_err(|_| Error::<T>::WithdrawFailed)?;
+		}
+
+		Self::delegation_withdraw(delegator, delegatee, value)
+	}
+
+	fn apply_slash(
+		delegatee: &Self::AccountId,
+		delegator: &Self::AccountId,
+		value: Self::Balance,
+		maybe_reporter: Option<Self::AccountId>,
+	) -> DispatchResult {
+		let mut delegation_register =
```

Comment:
```suggestion
		let mut ledger =
```

nit but this would be easier to read and make sense of for people familiar with the staking code.